src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -1448,6 +1492,118 @@ private function step_in_body() {
 		}
 	}
 
+	/**
+	 * Parses next element in the 'in select' insertion mode.
+	 *
+	 * This internal function performs the 'in select' insertion mode
+	 * logic for the generalized WP_HTML_Processor::step() function.
+	 *
+	 * @since 6.7.0
+	 *
+	 * @throws WP_HTML_Unsupported_Exception When encountering unsupported HTML input.
+	 *
+	 * @see https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inselect
+	 * @see WP_HTML_Processor::step
+	 *
+	 * @return bool Whether an element was found.
+	 */
+	private function step_in_select() {
+		$tag_name = $this->get_tag();
+		$op_sigil = $this->is_tag_closer() ? '-' : '+';
+		$op       = "{$op_sigil}{$tag_name}";
+
+		switch ( $op ) {
+			/*
+			 * > A start tag whose tag name is "html"
+			 */
+			case '+HTML':
+				$this->last_error = self::ERROR_UNSUPPORTED;
+				throw new WP_HTML_Unsupported_Exception( "Cannot process {$tag_name} element." );
+
+			/*
+			 * > A start tag whose tag name is "option"
+			 */
+			case '+OPTION':
+				$current_node = $this->state->stack_of_open_elements->current_node();
+				if ( $current_node && 'OPTION' === $current_node->node_name ) {
+					$this->state->stack_of_open_elements->pop();
+				}
+				$this->insert_html_element( $this->state->current_token );
+				return true;
+
+			/*
+			 * > A start tag whose tag name is "optgroup"
+			 * > A start tag whose tag name is "hr"
+			 */
+			case '+OPTGROUP':
+			case '+HR':
+				$current_node = $this->state->stack_of_open_elements->current_node();
+				if ( $current_node && 'OPTION' === $current_node->node_name ) {
+					$this->state->stack_of_open_elements->pop();
+					// If we've popped, update the current_node
+					$current_node = $this->state->stack_of_open_elements->current_node();
+				}
+
+				if ( $current_node && 'OPTGROUP' === $current_node->node_name ) {
+					$this->state->stack_of_open_elements->pop();
+				}
+
+				$this->insert_html_element( $this->state->current_token );
+				return true;
+
+			/*
+			 * > An end tag whose tag name is "optgroup"
+			 */
+			case '-OPTGROUP':
+				$walker       = $this->state->stack_of_open_elements->walk_up();
+				$current_node = $walker->current();
+				if ( ! $current_node ) {
+					return $this->step();
+				}
+				if ( 'OPTGROUP' === $current_node->node_name ) {
+					$this->state->stack_of_open_elements->pop();
+					return true;
+				}
+
+				$walker->next();
+				$current_node_parent = $walker->current();
```

Comment:
doesn't this section belong above the check for `OPTGROUP`?

Reply:
```php
$current_node = $this->state->stack_of_open_elements->current_node()
if ( 'OPTION' === $current_node ) {
	foreach ( $this->state->stack_of_open_elements->walk_up( $current_node ) as $parent ) {
		break;
	}
	if ( 'OPTGROUP' === $parent->node_name ) {
		$this->state->stack_of_open_elements->pop();
	}
}
```

Reply:
I simplified this, much better.